crates/red_knot_python_semantic/src/semantic_index/usedef.rs

```diff
@@ -0,0 +1,137 @@
+use crate::ast_node_ref::AstNodeRef;
+use crate::db::Db;
+use crate::semantic_index::ast_ids::ScopedUseId;
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::symbol::{FileScopeId, ScopedSymbolId};
+use ruff_db::files::File;
+use ruff_index::IndexVec;
+use ruff_python_ast as ast;
+use salsa;
+
+/// An independently type-inferable expression.
+///
+/// Includes constraint expressions (e.g. if tests) and the RHS of an unpacking assignment.
+#[salsa::tracked]
+pub(crate) struct Expression<'db> {
+    /// The file in which the expression occurs.
+    #[id]
+    pub(crate) file: File,
+
+    /// The scope in which the expression occurs.
+    #[id]
+    pub(crate) scope: FileScopeId,
```

Comment:
I think we also want the `ScopedExpressionId` as part of the `id`. Otherwise all expressions from the same scope have the same id.

Reply:
Hmm. Is this also a problem with our current representation of `Definition`? https://github.com/astral-sh/ruff/blob/main/crates/red_knot_python_semantic/src/semantic_index/definition.rs#L12-L21

There can be multiple definitions of the same symbol in the same scope, but the id of definition contains only file, scope, and symbol.

I'm a little fuzzy on the implications of creating multiple tracked structs that share an id, or whether we should really be using id in these structs at all. The description of it at https://salsa-rs.github.io/salsa/overview.html#id-fields is all about tracking identity across re-orderings, but I think any reordering of Definitions or Expressions in a scope is going to change symbol IDs and expression IDs anyway, so I'm not sure if there is any useful "identity" that we are tracking here at all.

Reply:
It depends on the query that creates the structure. `Definitions` are created as part of the `semantic_index` query. Adding or removing any definition would offset all definitions by one. The worst-case outcome is that all queries taking a definition as an argument now return a different result (worst-case invalidation). 

This is mitigated by including the `scope` and `symbol` in the ID. The scope ensures that adding a definition to an outer (or even inner) scope doesn't change the IDs of the current scope. Whether we need `symbol_id` and `file_id` is unclear to me. We could probably do without them. 

The fact that `symbol_id` isn't unique doesn't matter. If two structs have the same ID, Salsa relies on order (at least that's what I understand). 

I would need to have a closer look at `Expression` to understand what makes sense here. So yes, maybe scope itself is sufficient, because we then only depend on the order of expressions. But we might be able to do better by including an additional id so that adding / removing expressions doesn't invalidate all expressions in a scope

Reply:
Makes sense. The thing I don't understand about adding a `ScopedExpressionId` as part of the id here is that, given how we number expressions in-order within a scope, adding/removing expressions is going to result in changing the `ScopedExpressionId` of all the subsequent expressions anyway, right? So it's not clear to me how matching up by `ScopedExpressionId` is actually better (within a scope) than matching up by order of expressions; it seems like they are basically equivalent. In fact it seems like not having the `ScopedExpressionId` might be better, since only a few expressions are interned; without the `ScopedExpressionId` adding/removing non-interned expressions won't invalidate, but if we included the `ScopedExpressionId` then it would invalidate.